homeassistant/components/signal_messenger/notify.py

```diff
@@ -50,30 +53,78 @@ def __init__(self, recp_nrs, signal_cli_rest_api):
         self._signal_cli_rest_api = signal_cli_rest_api
 
     def send_message(self, message="", **kwargs):
-        """Send a message to a one or more recipients.
-
-        Additionally a file can be attached.
-        """
+        """Send a message to a one or more recipients. Additionally a file can be attached."""
 
         _LOGGER.debug("Sending signal message")
 
         data = kwargs.get(ATTR_DATA)
 
-        filenames = None
-        if data is not None:
-            if ATTR_FILENAMES in data:
-                filenames = data[ATTR_FILENAMES]
-            if ATTR_FILENAME in data:
-                _LOGGER.warning(
-                    "The 'attachment' option is deprecated, please replace it with 'attachments'. This option will become invalid in version 0.108"
-                )
-                if filenames is None:
-                    filenames = [data[ATTR_FILENAME]]
-                else:
-                    filenames.append(data[ATTR_FILENAME])
+        filenames = self.get_filenames(data)
+        attachments_as_bytes = self.get_attachments_as_bytes(
+            data, CONF_MAX_ALLOWED_DOWNLOAD_SIZE_BYTES
+        )
 
         try:
-            self._signal_cli_rest_api.send_message(message, self._recp_nrs, filenames)
+            self._signal_cli_rest_api.send_message(
+                message, self._recp_nrs, filenames, attachments_as_bytes
+            )
         except SignalCliRestApiError as ex:
             _LOGGER.error("%s", ex)
             raise ex
+
+    @staticmethod
+    def get_filenames(data):
+        """Extract attachment filenames from data."""
+        filenames = None
+
+        if data is None:
+            return None
+        if ATTR_FILENAMES in data:
+            filenames = data[ATTR_FILENAMES]
+        if ATTR_FILENAME in data:
+            _LOGGER.warning(
+                "The 'attachment' option is deprecated, please replace it with 'attachments'. "
+                "This option will become invalid in version 0.108"
+            )
+            if filenames is None:
+                filenames = [data[ATTR_FILENAME]]
+            else:
+                filenames.append(data[ATTR_FILENAME])
+
+        return filenames
+
+    @staticmethod
+    def get_attachments_as_bytes(data, attachment_size_limit):
+        """Retrieve attachments from URLs defined in data."""
+        attachments_as_bytes = []
+
+        if data is None or ATTR_URLS not in data:
+            return attachments_as_bytes
+
+        urls = data[ATTR_URLS]
```

Comment:
Can we also verify that this is a list?

Reply:
Type check added in d9415bcc03823e892c547210f4faf4435934fbfc